src/DicomWebClient/API/IWadoService.cs

```diff
@@ -0,0 +1,150 @@
+            params DicomTransferSyntax[] transferSyntaxes);
+
+        /// <summary>
+        /// Retrieves the metadata of all instances specified in the study.
+        /// </summary>
+        /// <typeparam name="T">T must be type of <see cref="DicomDataset"/> or <see cref="string"/></typeparam>
+        /// <param name="studyInstanceUid">Study Instance UID</param>
+        /// <param name="seriesInstanceUid">Series Instance UID</param>
+        /// <returns>An enumerable of <c>T</c> containing DICOM instances for the series.</returns>
+        IAsyncEnumerable<T> RetrieveMetadata<T>(
+            string studyInstanceUid,
+            string seriesInstanceUid);
+
+        /// <summary>
+        /// Retrieves a DICOM instance.
+        /// </summary>
+        /// <param name="studyInstanceUid">Study Instance UID</param>
+        /// <param name="seriesInstanceUid">Series Instance UID</param>
+        /// <param name="sopInstanceUid">SOP Instance UID</param>
+        /// <param name="transferSyntaxes">An array of supported transfer syntaxes. Default set to Explicit VR Little Endian (1.2.840.10008.1.2.1)</param>
+        /// <returns>A <see cref="DicomFile"/> representing the DICOM instance.</returns>
+        Task<DicomFile> Retrieve(
+            string studyInstanceUid,
+            string seriesInstanceUid,
+            string sopInstanceUid,
+            params DicomTransferSyntax[] transferSyntaxes);
+
+        /// <summary>
+        /// Retrieves the metadata of the specified DICOM instance.
+        /// </summary>
+        /// <typeparam name="T">T must be type of <see cref="DicomDataset"/> or <see cref="string"/></typeparam>
+        /// <param name="studyInstanceUid">Study Instance UID</param>
+        /// <param name="seriesInstanceUid">Series Instance UID</param>
+        /// <param name="sopInstanceUid">SOP Instance UID</param>
+        /// <returns>A <c>T</c> containing DICOM metadata for the instance.</returns>
+        Task<T> RetrieveMetadata<T>(
+            string studyInstanceUid,
+            string seriesInstanceUid,
+            string sopInstanceUid);
+
+        /// <summary>
+        /// Retrieves one or more frames from a multi-frame DICOM instance.
+        /// </summary>
+        /// <param name="studyInstanceUid">Study Instance UID</param>
+        /// <param name="seriesInstanceUid">Series Instance UID</param>
+        /// <param name="sopInstanceUid">SOP Instance UID</param>
+        /// <param name="frameNumbers">The frames to retrieve within a multi-frame instance. (One-based indices)</param>
+        /// <param name="transferSyntaxes">An array of supported transfer syntaxes. Default set to Explicit VR Little Endian (1.2.840.10008.1.2.1)</param>
+        /// <returns>A <see cref="DicomFile"/> representing the DICOM instance.</returns>
+        Task<DicomFile> Retrieve(
+            string studyInstanceUid,
+            string seriesInstanceUid,
+            string sopInstanceUid,
+            uint[] frameNumbers,
```

Comment:
nit: just for reference `IReadOnlyList<uint>` is significantly more flexible than `uint[]` and it offers psudo-const protection.